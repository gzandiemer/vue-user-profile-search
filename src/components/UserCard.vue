<template>
  <div>
    <router-link :to="`/user/${user.login.uuid}`">
      <img
        :src="user.picture.thumbnail"
        alt="User picture"
      >
      <h2>{{ formattedName }}</h2>
      <p>{{ user.email }}</p>
    </router-link>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
export default defineComponent({
  props: {
    user: {
      type: Object,
      required: true
    }
  },
  computed: {
    formattedName() {
      const first =
        this.user.name.first.charAt(0).toUpperCase() +
        this.user.name.first.slice(1);
      const last =
        this.user.name.last.charAt(0).toUpperCase() +
        this.user.name.last.slice(1);
      return `${first} ${last}`;
    }
  }
});
</script>

<style></style>
