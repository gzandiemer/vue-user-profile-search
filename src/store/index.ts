import { createStore } from 'vuex';
import { User } from '../types';

interface State {
  users: User[];
  page: number;
}

export default createStore({
  state: {
    users: [],
    page: 1,
  } as State,
  mutations: {
    setUsers(state: State, users: User[]) {
      state.users = state.users.concat(users);
    },
    incrementPage(state: State){
      state.page ++;
    }
  },
  actions: {
    async fetchUsers({ commit, state }) {
      const response = await fetch(`https://randomuser.me/api/?results=25&page=${state.page}`);
      const data = await response.json();
      commit('setUsers', data.results);
      commit('incrementPage');
    }
  },
  getters: {
    getUser: (state: State) => (id: string) => {
      return state.users.find((user) => user.login.uuid === id);
    }
  }
});

