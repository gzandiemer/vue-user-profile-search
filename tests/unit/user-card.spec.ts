import { mount } from "@vue/test-utils";
import UserCard from "@/components/UserCard.vue";
import { createStore } from "vuex";
import { cloneDeep } from 'lodash';
import { State } from "@/store";

interface MyStoreOptions {
  state: State;
  actions: any;
  mutations: any;
  getters: any;
}

describe('UserCard.vue', () => {
  let state: State;
  let actions: any;
  let mutations: any;
  let getters: any;
  let localStore: any;
  let localStoreConfig: MyStoreOptions;
  let wrapper: any;

  const mockUser = {
    name: {
      first: 'Test',
      last: 'User'
    },
    email: 'test.user@example.com',
    picture: {
      thumbnail: 'https://via.placeholder.com/150'
    }
  };

  beforeEach(() => {
    state = {
      users: [],
      selectedUser: null,
      searchQuery: '',
      filter: '',
      page: 1,
      gender: ''
    };

    actions = {};

    mutations = {
      setSelectedUser: jest.fn()
    };

    getters = {};

    localStoreConfig = {
      state,
      actions,
      mutations,
      getters
    };

    localStore = createStore(cloneDeep(localStoreConfig));
  });

  beforeEach(() => {
    wrapper = mount(UserCard, { 
      props: { user: mockUser },
      global: { plugins: [localStore] } 
    });
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays the user name and email', () => {
    const userInfo = wrapper.find('.user-info');
    expect(userInfo.text()).toContain('Test User');
    expect(userInfo.text()).toContain('test.user@example.com');
  });

  it('emits "user-selected" event when user card is clicked', async () => {
    await wrapper.find('.user-card').trigger('click');
    expect(wrapper.emitted('user-selected')).toBeTruthy();
    expect(wrapper.emitted('user-selected')?.[0]).toEqual([mockUser]);
  });
});
