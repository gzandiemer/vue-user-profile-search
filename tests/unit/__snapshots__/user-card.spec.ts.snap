// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard.vue should render correctly 1`] = `
<div class="user-card"><img class="user-picture" src="https://via.placeholder.com/150" alt="User picture">
  <div class="user-info">
    <h2>Test User</h2>
    <p>test.user@example.com</p>
  </div>
</div>
`;
